src/Npgsql/NpgsqlBatchCommandCollection.cs

```diff
@@ -97,8 +99,16 @@ protected override void SetBatchCommand(int index, DbBatchCommand batchCommand)
         => _list[index] = Cast(batchCommand);
 
     static NpgsqlBatchCommand Cast(DbBatchCommand? value)
-        => value is NpgsqlBatchCommand c
-            ? c
-            : throw new InvalidCastException(
-                $"The value \"{value}\" is not of type \"{nameof(NpgsqlBatchCommand)}\" and cannot be used in this batch command collection.");
+    {
+        var castedValue = value as NpgsqlBatchCommand;
+        if (castedValue is null)
+            ThrowInvalidCastException(value);
+
+        return castedValue;
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining), DoesNotReturn]
```

Comment:
```suggestion
    [DoesNotReturn]
```

The "no inlining" should be removed in order to allow to JIT to see what the method does. The JIT then sees that it throws, so won't return and thus can produce better code by moving that piece to a cold section and won't inline it anyway.

With "no inlining" the JIT doesn't look into the method, so it can't do any optimizations around that.

Reply:
Yeah, you're right. Tested this via a [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzNNdvHgwCyACisYGC7JI4wAlMdPnfvAGYMdi5uMAwAhAC8DHQePr4JDAAqABZQEADuAPIcGFn+bNhcAOYwdh4a1IlmxADszq7ulb4AvvEM7ToMjrakDtYh7l4mVdUBQYNhUTFxo9XmqenZufmFJTB9Fe2+dQ2hzeZtcx3HXcQoyWmZOXkFRaXl7SPzGFcZDFww7wCCUMUcAh4N1W9xgAFEECoxHwIFxygczEdfO0ANo2GCvCAAEwAkuJJHZ0ZjcfistDeLDcCwAHIQHFcSRWKzFDwAXU6unOlyWwLu602T225leSw+XwYv3+gLyKz5pQhUJhcK2xyOLSAA==) and it indeed produces a better code. Thanks!

Reply:
As additional note / history: I think the JIT got taught in .NET 5 how to handle this. Before it was best practice to apply "no inlining" attribute to give the JIT the needed hint.  Now, and especially with dynamic profile guided optimizations, it's counter-productive to use that attribute for throw-only-methods.